tests/test_interconnect/test_interconnect_cgra.py

```diff
@@ -419,3 +420,125 @@ def test_interconnect_sram(cw_files, add_pd, io_sides):
+def test_interconnect_fifo(cw_files, add_pd, io_sides, depth):
+    chip_size = 2
+    interconnect = create_cgra(chip_size, chip_size, io_sides,
+                               num_tracks=3,
+                               add_pd=add_pd,
+                               mem_ratio=(1, 2))
+
+    netlist = {
+        "e0": [("I0", "io2f_16"), ("m0", "data_in")],
+        "e1": [("i3", "io2f_1"), ("m0", "wen_in")],
+        "e2": [("i4", "io2f_1"), ("m0", "ren_in")],
+        "e3": [("m0", "data_out"), ("I1", "f2io_16")],
+        "e4": [("m0", "valid_out"), ("i4", "f2io_1")],
+        "e5": [("m0", "empty"), ("i2", "f2io_1")],
+        "e6": [("m0", "full"), ("i3", "f2io_1")]
+    }
+    bus = {"e0": 16, "e1": 1, "e2": 1, "e3": 16, "e4": 1, "e5": 1, "e6": 1}
+
+    placement, routing = pnr(interconnect, (netlist, bus))
+    config_data = interconnect.get_route_bitstream(routing)
+
+    # in this case we configure m0 as fifo mode
+    mode = Mode.FIFO
+    tile_en = 1
+
+    mem_x, mem_y = placement["m0"]
+    memtile = interconnect.tile_circuits[(mem_x, mem_y)]
+    mcore = memtile.core
+    config_data.append((interconnect.get_config_addr(
+                        mcore.get_reg_index("depth"),
+                        0, mem_x, mem_y), depth))
+    config_data.append((interconnect.get_config_addr(
+                        mcore.get_reg_index("mode"),
+                        0, mem_x, mem_y), mode.value))
+    config_data.append((interconnect.get_config_addr(
+                        mcore.get_reg_index("tile_en"),
+                        0, mem_x, mem_y), tile_en))
+    config_data.append((interconnect.get_config_addr(
+                        mcore.get_reg_index("almost_count"),
+                        0, mem_x, mem_y), 3))
+
+    circuit = interconnect.circuit()
+
+    tester = BasicTester(circuit, circuit.clk, circuit.reset)
+    tester.reset()
+    for addr, index in config_data:
+        tester.configure(addr, index)
+        tester.config_read(addr)
+        tester.eval()
+        tester.expect(circuit.read_config_data, index)
+
+    src_x, src_y = placement["I0"]
+    src = f"glb2io_16_X{src_x:02X}_Y{src_y:02X}"
+    dst_x, dst_y = placement["I1"]
+    dst = f"io2glb_16_X{dst_x:02X}_Y{dst_y:02X}"
+    wen_x, wen_y = placement["i3"]
+    wen = f"glb2io_1_X{wen_x:02X}_Y{wen_y:02X}"
+    valid_x, valid_y = placement["i4"]
+    valid = f"io2glb_1_X{valid_x:02X}_Y{valid_y:02X}"
+    ren_x, ren_y = placement["i4"]
+    ren = f"glb2io_1_X{ren_x:02X}_Y{ren_y:02X}"
+    full_x, full_y = placement["i3"]
+    full = f"io2glb_1_X{full_x:02X}_Y{full_y:02X}"
+    empty_x, empty_y = placement["i2"]
+    empty = f"io2glb_1_X{empty_x:02X}_Y{empty_y:02X}"
+
+    fifo = deque()
+    for i in range(1024):
+
+        tester.expect(circuit.interface[empty], len(fifo) == 0)
+        tester.expect(circuit.interface[full], len(fifo) == depth)
+        # Valid is just ~empty
+        tester.expect(circuit.interface[valid], len(fifo) > 0)
```

Comment:
Shouldn't `valid` also depend on `circuit.interface[ren]`?

Reply:
So it could indicate that there is data to be read or that the data on the line is good (basically once it goes high, it would only go low again if you tried to read when it was empty)

Reply:
made the valid correct in the last commit